src/Control/Effect/Class.hs

```diff
@@ -11,78 +11,102 @@ module Control.Effect.Class
 ) where
 
 import Data.Coerce
+import Data.Functor.Identity
+import Data.Kind (Constraint)
 import GHC.Generics
 
 -- | The class of effect types, which must:
 --
 --   1. Be functorial in their last two arguments, and
---   2. Support threading effects in higher-order positions through using the carrier’s suspended state.
+--   2. Support threading effects in higher-order positions through using the carrier’s suspended context.
+--
+-- Effects may additionally constrain the types of context that they support threading through their actions by defining an instance of the associated 'CanHandle' type family. If no definition is given, it defaults to 'Functor'.
 --
 -- All first-order effects (those without existential occurrences of @m@) admit a default definition of 'handle' provided a 'Generic1' instance is available for the effect.
 --
 -- @since 1.0.0.0
-class Effect sig where
-  -- | Handle any effects in a signature by threading the carrier’s state all the way through to the continuation.
-  handle :: (Functor f, Monad m)
-         => f ()
-         -> (forall x . f (m x) -> n (f x))
-         -> sig m a
-         -> sig n (f a)
-  default handle :: (Functor f, Monad m, Generic1 (sig m), Generic1 (sig n), GEffect m n (Rep1 (sig m)) (Rep1 (sig n)))
-                 => f ()
-                 -> (forall x . f (m x) -> n (f x))
-                 -> sig m a
-                 -> sig n (f a)
+class CanHandle sig Identity => Effect sig where
+  -- | Constrain the type of context algebras can pass to 'handle'.
+  --
+  -- Defaults to 'Functor'. Some effects may require a more restrictive constraint to thread handlers through. It is recommended, but not enforced, that imposed constraints be subclasses of 'Functor' wherever possible to ensure compatibility with the broadest variety of algebras.
+  type CanHandle sig (ctx :: (* -> *)) :: Constraint
+  type CanHandle sig ctx = Functor ctx
+
+  -- | Handle any effects in a signature by threading the algebra’s handler all the way through to the continuation, starting from some initial context.
+  --
+  -- The handler is required to adhere to the following laws:
+  --
+  -- @
+  -- handler . 'fmap' 'pure' = 'pure'
+  -- @
+  -- @
+  -- handler . 'fmap' (k '=<<') = handler . 'fmap' k 'Control.Monad.<=<' handler
+  -- @
+  --
+  -- respectively expressing that the handler does not alter the context of pure computations, and that the handler distributes over monadic composition.
+  handle
+    :: (Monad m, Monad n, CanHandle sig ctx)
+    => ctx ()                              -- ^ The initial context.
+    -> (forall x . ctx (m x) -> n (ctx x)) -- ^ A handler for actions in a context, producing actions with a derived context.
+    -> sig m a                             -- ^ The effect to thread the handler through.
+    -> sig n (ctx a)
```

Comment:
1. We’ve added a `Monad n` constraint. This is technically backwards-incompatible for callers of `handle`, but I doubt anybody’s actually using these for non-`Monad`s (remember that `Algebra`, basically the only thing that ever calls this, has `Monad` as a superclass), so I seriously doubt it will ever impact anyone. This was done both to follow the literature (note that the laws assume `n` is a `Monad`) and to support effects embedding actions producing values of concrete types. (cf `Thread` in the tests.)

2. I’ve renamed the `f` parameter to `ctx` following the discussion of the name above.

3. Docs, obvs. I’m really leaning into this “thread” verbage. The literature uses it, and in my opinion it’s exactly the right metaphor, and I think the potential for confusion with e.g. concurrency is minimized both by context and usage as a verb. In #300, I’ve actually renamed `handle` to `thread`, but I wanted to review that change disjointly from the rest of this PR.